### ISSUE TITLE
Design: Insurgency vs Linear progression

### ISSUE THREAD

**Issue by thebgpikester** [KaukausInsurgency/kaukasusinsurgency-dcs#204]
After some time looking under the hood and play testing it occurred to me there is a limit in the design of Insurgency ( it is normal to have limits) but this limit perhaps prevents a larger scope that other players might like to try to change or challenge.

[SPOILERS]
The current method of zone control doesn't provide any real connection from zone to zone. Because it's an insurgency, its OK for enemy units to pop up behind enemy lines, This will continue, and the game becomes a type of "Whack-a-mole".

That's perfectly acceptable, after all, Insurgency is exactly that!

So my "Issue" point is that, there is a larger picture here on design that is perhaps missed. It might be labelled as "scope creep" but sooner or later, people want to try traditional warfare where supply lines and "Conquest" is achievable as an end game state. This type of persistent game is also easier for the average player to comprehend and work towards.

What's I'm proposing is that zones become linked in terms of enemy action. IE Capture this, friendly forces reinforce, enemy forces can no longer spawn near it. IE, decouple the link between the enemy spawns and the target zone, but link it instead to the zone that enemy control.

This is best painted in pictures, hard to describe: 
## An example;
4 zones, A, B, C, D
Friendly forces start in A and proceed to dominate all zones.
Enemy forces start in B and attempt to capture A
Friendly forces capture B
Enemy forces start in C and try to capture B (BUT CRUCIALLY, NOT "A")
Friendly forces try to capture C from a location close to B.
A>B>C>D

## Issues I can see:
-Each zone still has a relevant spawning enemy to try to capture it. BUT, you/someone would need to code the closure of the zones and prevent that. It's a fundamental shift and adds complexity.
- The ME does not know the progression and progression may not be linear. Consider A>B1 OR B2>C1 OR C2>D or even more if multiple "routes" are considered.

## Potential solution 
I've done this before, linear is pretty easy to do, forks are harder.
It goes like this and I'm attempting to already add it to my version of your framework:

You first collect all the zone and airbase ownerships on a schedule. Inside the schedule is a giant set of IF statements that govern Spawns, The "secret sauce" is creating spawns first, setting them to "SpawnScheduled()" with an InitLimit() that sets the maximum amount in play at one time and then immediately hitting "SpawnScheduleStop()".
The Schedule runs, and based on the results of the zone ownership, you start or stop spawns. The trick here is to stop the right ones.

Let's use a Georgia map example in semi pseudo code.
SCHEDULE every 10 seconds(
If Batumi is blue and Kobuleti is red then
SpawnScheduleStart() Blue Batumi group going to Kobuleti
SpawnScheduleStart() Red Kobuleti group going to Batumi
SpawnScheduleStop() all other spawns except side missions
end
If Batumi is blue and Kobuleti is blue and Kutaisi is red then
SpawnScheduleStart() Blue Kobuleti group going to Kutaisi
SpawnScheduleStart() Red Kutaisi group going to Kobuleti
stop all the other spawns
end
etc
SCHEDULE

It gets complex but have a think, see if it opens any ideas for discussion.

**Comment by thebgpikester** [comment]
I considered a way to manage campaign progression using the 1,2,3,6,12,24,48 sequence. (I'm neither a mathematician or a coder i'm afraid!) If each zone has one of these sequences, then every combination is a unique number. if the sequence is Batumi, kobuleti, Senaki, kutaisi then owning Senaki and kutaisi = 9 and is a unique number that you can use, same as 11 can easily been seen as owning Kutaisi, Senaki and Kobuleti. Which would simplify the "IF" statement. But i'm not sure if it helps making the code any easier to make generic. I'm rambling and will shut up!

**Comment by Igneous01** [comment]
It's an interesting idea, would require creating a graph or network of nodes to map out the decision making in the AI. Whether this is something the user configured or is auto created is a different story.

Ie. if you have 7 zones

![image](https://user-images.githubusercontent.com/5571589/35769173-e17bc5b6-08d4-11e8-972e-1a0d52722265.png)

You can create some heuristic algorithm that checks which zones are controlled by the enemy, then traverse the neighboring nodes connected to it and find nodes which are neutral or controlled by the enemy team. Then design some cost function that maybe checks things like distance, strength of forces, neutral/friendly to determine what the cost would be to attack the zone. Then once all the zones/moves are calculated, have the enemy choose the top x zones to attack that have the cheapest cost associated. 
Would need to able to discard duplicate nodes (ie if zones D and B are controlled by the enemy, having both of them spawn units to attack C would need to be mitigated to avoid crazy amounts of spawning on the same point)

**Comment by thebgpikester** [comment]
Yep, Ideally avoiding duplication of something like Autogft at https://github.com/birgersp/dcs-autogft which already does this but without major costing of zones (not sure how Birger chooses the zone),

However, the strong point on your solution is the short distances to travel. If it begins that Ground forces travel from node to node, it could get expensive in CPU. You have to decide, what is in scope and what isn't, since, having tested scripting and spawning over the top of your framework, I know it works. I don't know how you rename groups between your "saves", but potentially any solution has to manage a restart in between. Hence there is a "scanning" method in AutoGFT that takes control of groups by prefix. I dont know if Autogft ever solved the scanning then renaming to something else.

**Comment by Igneous01** [comment]
Yeah I'm aware of gft, can probably crack it open and look at what he's done and see how much of that can be carried over.

Waypoints are saved during restarts because I ran into a similar problem with convoy resupplies. Currently the group name and waypoint are saved to file in a table, when the group is respawned in mission restart, it checks to see if it had a waypoint in the table. If it did it checks the distance of that groups current position after being spawned and the waypoint, if it's below some minimum distance, the waypoint task is ignored. If it's above the minimum distance, the group is reassigned the move waypoint task again.

**Comment by thebgpikester** [comment]
Since we are doing something manually over the top of KI, close to this, at some point when you are close to done, I can go through our attempt and you can take some inspiration. But for this reason we have also played down the use of the AICOM which is a major part of your  work. Not because we don't think it does things well, but because we are looking for finer control with a conventional battle.
